### PR TITLE
Implement reference to non-existent group regex rule

### PR DESCRIPTION
**Pull request by diokey**
This PR adds a new lint rule that catches references to a non-existent capture group.

To achieve this, I had to add a new AST node `RegExpNumericBackReference` and take into account octal escapes while parsing.

The backreference will match the group capture if it exists, otherwise, it will be treated as an octal escape char. 

For example 
```js
    String.raw`let foo = /([abc]+)=\1/;foo;`, // matches first capture group
    String.raw`let foo = /([abc]+)=\2/;foo;`, // matches \2 escaped
    String.raw`let foo = /([abc]+)=\8/;foo;`, // matches '8'
    String.raw`let foo = /([abc]+)=\9/;foo;`, // matches '9'
    String.raw`let foo = /([abc]+)=\119/;foo;`, // matches '\t' followd by '9'
    String.raw`let foo = /([abc]+)=\338/;foo;`, // matches \33 (char code 255) followed by '8'
    String.raw`let foo = /([abc]+)=\377/;foo;`, // matches \377 (char code 255)
    String.raw`let foo = /([abc]+)=\777/;foo;`, // matches \77  (char code 63) followed by '7'
```
The PR catches cases where the backreference is not an octal escape and does not refer to any existing group capture.

```js
    String.raw`let foo = /([abc]+)=\18/;foo;`, 
    String.raw`let foo = /([abc]+)=\49/;foo;`,
    String.raw`let foo = /([abc]+)=\78/;foo;`,
    String.raw`let foo = /([abc]+)=\99/;foo;`,
    String.raw`let foo = /(([abc])\19)+=\28/;foo;`,
```

image

I am no expert in either compilers or regex, so let me know if this is not the right approach. I used https://www.regular-expressions.info/backref.html and tested those example on https://regexr.com/ 

Thank you

### CODE REVIEW

File: packages/@romejs/codec-js-regexp/index.ts [hunk 1]
```diff
@@ -214,7 +226,7 @@ export const createRegExpParser = createParser((ParserCore) =>
 
           case 'x':
             {
-              const possibleHex = input.slice(get0(index) + 1, 2);
+              const possibleHex = input.slice(get0(index) + 1, get0(index) + 2);
```
Comment (diokey) [inline packages/@romejs/codec-js-regexp/index.ts]: @sebmck this is not related, but it like a wrong index. The condition below is always false. Maybe it should be `get0(index) + 2`

Comment (diokey) [reply packages/@romejs/codec-js-regexp/index.ts]: remove that change, but wanted to bring it to your attention

Comment (sebmck) [reply packages/@romejs/codec-js-regexp/index.ts]: Aha yeah it should be. Feel free to leave that in change in! Thanks for flagging.

File: packages/@romejs/codec-js-regexp/index.ts [hunk 2]
```diff
@@ -482,6 +533,35 @@ export const createRegExpParser = createParser((ParserCore) =>
...
+
+        // octal escapes
+        if (isOct(String(value))) {
+          const octal = parseInt(String(value), 8);
+          return {
+            type: 'RegExpCharacter',
+            value: String.fromCharCode(octal),
+            loc: this.finishLocFromToken(token),
+          };
+        }
```
Comment (sebmck) [inline packages/@romejs/codec-js-regexp/index.ts]: Is there a way we could put this into `tokenize` instead?

File: packages/@romejs/codec-js-regexp/index.ts (hunk 1 above)
Comment (diokey) [reply packages/@romejs/codec-js-regexp/index.ts]: Great, will do

File: packages/@romejs/codec-js-regexp/index.ts (hunk 2 above)
Comment (diokey) [reply packages/@romejs/codec-js-regexp/index.ts]: Sure, will do

File: packages/@romejs/codec-js-regexp/index.ts (hunk 1 above)
Comment (diokey) [reply packages/@romejs/codec-js-regexp/index.ts]: Done

File: packages/@romejs/codec-js-regexp/index.ts (hunk 2 above)
Comment (diokey) [reply packages/@romejs/codec-js-regexp/index.ts]: Done

**Comment by diokey** [comment]
@sebmck the CI is failing, but it doesn't look like it's related to this PR. Any ideas?

**Comment by sebmck** [comment]
This is awesome! Thank you so much!

**Pull request merged**